pkg/cli/image/extract/extract.go

```diff
@@ -357,7 +369,51 @@ func (o *ExtractOptions) Run() error {
 					return err
 				}
 
-				imageConfig, layers, err := imagemanifest.ManifestToImageConfig(ctx, srcManifest, repo.Blobs(ctx), location)
+				manifests, err := repo.Manifests(ctx)
+				if err != nil {
+					return err
+				}
+
+				// o.PlatformSpec is only set here from 'oc adm release new', not with 'oc image extract'.
+				if len(o.PlatformSpec) == 0 {
+					// If wildcard filter was passed and hasn't been processed, and there is more than one manifest, error
+					if o.FilterOptions.IsWildcardFilter() {
+						switch t := srcManifest.(type) {
+						case *manifestlist.DeserializedManifestList:
+							if len(t.Manifests) > 1 {
+								srcRef, err := imagesource.ParseReference(mapping.Image)
+								if err != nil {
+									return err
+								}
+								retriever := &info.ImageRetriever{
+									Image: map[string]imagesource.TypedImageReference{
+										mapping.Image: srcRef,
+									},
+									FileDir:         o.FileDir,
+									SecurityOptions: o.SecurityOptions,
+									ManifestListCallback: func(from string, list *manifestlist.DeserializedManifestList, all map[digest.Digest]distribution.Manifest) (map[digest.Digest]distribution.Manifest, error) {
+
+										buf := &bytes.Buffer{}
+										w := tabwriter.NewWriter(buf, 0, 0, 1, ' ', 0)
+										fmt.Fprintf(w, "  OS\tDIGEST\n")
+										for _, manifest := range list.Manifests {
+											fmt.Fprintf(w, "  %s\t%s\n", imagemanifest.PlatformSpecString(manifest.Platform), manifest.Digest)
+										}
+										w.Flush()
+										return nil, fmt.Errorf("the image is a manifest list and contains multiple images - use --filter-by-os to select from:\n\n%s\nor, to default to system '%s/%s' don't pass --filter-by-os.\n\n", buf.String(), runtime.GOOS, runtime.GOARCH)
+									},
+								}
+								if err := retriever.Run(); err != nil {
+									return err
+								}
+							}
+							// Setting PlatformSpec here enables better error message if given --filter-by-os os/arch does not match an
+							// available os/arch when srcManifest is a manifest list and imagemanifest.ManifestToImageConfig errors.
+							o.PlatformSpec = o.FilterOptions.FilterByOS
+						}
+					}
+				}
+				imageConfig, layers, err := imagemanifest.ManifestToImageConfig(ctx, srcManifest, repo.Blobs(ctx), manifests, location, o.PlatformSpec)
```

Comment:
@sallyom so what I see there:
* imagemanifest.FirstManifest filters manifest lists and returns `srcManifest`: either a manifest for the selected platform or a manifest list
* if PlatformSpec is not set and FilterOptions is a wildcard filter and `srcManifest` is a manifest list:
    * if `srcManifest` has more than one manifest, print an error and exit
    * otherwise set PlatformSpec equal to FilterByOS
* if we still have a manifest list, ManifestToImageConfig is going to filter it again, but using PlatformSpec

Can you incorporate filtering by PlatformSpec into FirstManifest? I think you can update FirstManifest and rename it into, let's say, PlatformManifest. It's called from two places and both places don't want it to return a manifest list.

Reply:
I will try, thanks!

Reply:
ptal, updated, thanks